### PR TITLE
Swap to HTTP cache

### PR DESCRIPTION
**Pull request by bovine3dom**
~~This improves runtime performance for non-comparisons (although that could be improved too if we felt like it) in exchange for a bit of jank at first load.~~

~~I'm wondering whether it's worth turning `getData` into a web worker that does fetching, caching and warmup all on its own to prevent jank in the main thread.~~

~~(Where "jank" means moving the map around is a bit slow).~~

### CODE REVIEW

**Comment by bovine3dom** [comment]
https://github.com/IntegratedTransportPlanning/gcvt/commit/b94a514119cf08035817d9160de23d17351cbdb3#comments

@cmcaine this switches us to a HTTP cache (and tells CDNs etc that they can cache us). It works really well.

I'm hesitant to merge it until we're closer to done, though, because unlike the JS cache it's a bit fiddlier to bypass. I can foresee us getting confused because bugs we've fixed haven't been fixed.

**Comment by cmcaine** [comment]
My plan was that we add a UUID to the URL so that different paths are requested if the source is changed.

So there's some API endpoint that you can ask for the current UUID, and then you add that to all your requests or something?

We'd need to add something to getData so if it fails to get a resource it tries to update the UUID and tries again or something?

**Comment by bovine3dom** [comment]
Daft solution: stick an ID at the top of the Julia file which we change manually. Essentially a version number.

In JS, our first task is to ask Julia what the current ID on a route which gives us a special "never cache me" response. We add this ID to the query string, which Julia just ignores.

Edit: we'd probably need `memoize` to return so that we can cache the ID inside `getData` without having to think about it too much.

We shouldn't need to do anything else.

**Comment by bovine3dom** [comment]
Works a treat.

File: src/frontend/js/index.js [hunk 1]
```diff
@@ -96,7 +98,17 @@ const nerf = x => (1+erf(x*2-1))/2
...
+
+async function _getApiVersion() {
+    try {
+        return (await (await fetch("/api/version")).json())["version"]
+    } catch (e) {
+        console.error(`Error getting data from:\n/api/version\n\n`, e)
+    }
+}
+
+const getApiVersion = memoize(_getApiVersion)
```
Comment (cmcaine) [inline src/frontend/js/index.js]: Seems a bit silly to use memoize for a function with no args?

`const API_VERSION_PROMISE = getApiVersion()` would be okay, I think?

File: src/frontend/js/index.js [hunk 2]
```diff
@@ -96,7 +98,17 @@ const nerf = x => (1+erf(x*2-1))/2
 
 // get data from Julia:
 const getData = async endpoint =>
-    (await fetch("/api/" + endpoint)).json().catch(e => console.error(`Error getting data from:\n/api/${endpoint}\n\n`, e))
+    (await fetch("/api/" + endpoint + (endpoint.includes("?") ? "&" : "?") + "v=" + (await getApiVersion()))).json().catch(e => console.error(`Error getting data from:\n/api/${endpoint}\n\n`, e))
```
Comment (cmcaine) [inline src/frontend/js/index.js]: This looks like a mistake? You do .json() in getApiVersion, so you shouldn't need to do it here as well.

Comment (bovine3dom) [reply src/frontend/js/index.js]: It's `fetch(...).json()` : )

File: src/frontend/js/index.js (hunk 1 above)
Comment (bovine3dom) [reply src/frontend/js/index.js]: Yes, it is silly. Ta

**Review by cmcaine — APPROVED** [review]

**Pull request merged**